internal/profiling/profiling.go

```diff
@@ -0,0 +1,252 @@
+// lead to data races.
+type Stat struct {
+	// Each Stat has a tag associated with it to identify it in post-processing.
+	// This is can be any arbitrary string.
+	StatTag string
+	// Stats may also need to store other unstructured information specific to
+	// this stat. For example, a StreamStat will use these bytes to encode the
+	// connection number and the stream ID for each RPC, thereby uniquely
+	// identifying it. The underlying encoding for this is unspecified.
+	Metadata []byte
+	// A collection of Timers and a mutex for append operations on the slice.
+	mu     sync.Mutex
+	Timers []Timer
+}
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation. A typical
+// unary RPC creates 80-100 timers for various things. While this number is
+// purely anecdotal and may change in the future as the resolution of profiling
+// increases or decreases, it serves as a good estimate for what the initial
+// allocation size should be.
+const defaultStatAllocatedTimers int32 = 128
+
+// NewStat creates and returns a new Stat object.
+func NewStat(statTag string) *Stat {
+	return &Stat{
+		StatTag: statTag,
+		Timers:  make([]Timer, 0, defaultStatAllocatedTimers),
+	}
+}
+
+// NewTimer creates a Timer object within the given stat if stat is non-nil.
+// The value passed in timerTag will be attached to the newly created Timer.
+// NewTimer also automatically sets the Begin value of the Timer to the current
+// time.
+//
+// The user is expected to call stat.Egress with the returned index as argument
+// to mark the end. The return value is not a pointer to a Timer object because
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
+// new timer every time. Instead a slice of timers large enough to hold most
+// data is created and elements from this slice are used. This should put less
+// pressure on the GC too.
+func (stat *Stat) NewTimer(timerTag string) int {
+	if stat == nil {
+		return 0
+	}
+
+	// TODO(adtac): Make this lock-free? We don't expect much contention on
+	// this, so I don't predict a significant improvement in performance. Using
+	// a lock-free, concurrent queue will require us to write our own
+	// reallocation and copy, which is probably unnecessary code complexity if
+	// the improvement isn't significant.
+	stat.mu.Lock()
+	stat.Timers = append(stat.Timers, Timer{
+		TimerTag: timerTag,
+		Begin:    time.Now(),
+		GoID:     goid(),
+	})
+	index := len(stat.Timers) - 1
+	stat.mu.Unlock()
+	return index
+}
+
+// Egress marks the completion of a given timer within a stat.
+func (stat *Stat) Egress(index int) {
+	if stat == nil {
+		return
+	}
+
+	stat.mu.Lock()
+	if index < len(stat.Timers) {
+		stat.Timers[index].End = time.Now()
```

Comment:
Should we call time.Now() before taking the lock?

Reply:
Yes, good point. While there is a case to be made that this might call time.Now wastefully even if the index check fails, that should ideally never happen. Done.